core/src/main/java/org/apache/calcite/util/NameMultimap.java

```diff
@@ -87,35 +88,34 @@ public NameMultimap(boolean allowsDuplicates) {
    * existing entry will be overwritten.
    */
   public void put(String name, V v) {
-    List<V> list = map().computeIfAbsent(name, k -> new ArrayList<>());
-    int index = -1;
-    if (!allowsDuplicates) {
-      index = list.indexOf(v);
-    }
-    if (index >= 0) {
-      list.set(index, v);
+    Collection<V> collection = null;
+    if (allowsDuplicates) {
+      collection = map().computeIfAbsent(name, k -> new ArrayList<>());
     } else {
-      list.add(v);
+      collection = map().computeIfAbsent(name, k -> new HashSet<>());
+      // Remove old instance so it gets overridden.
+      collection.remove(v);
```

Comment:
Why do we need to explicitly remove here? Shouldn't the HashSet automatically handle the overwriting?

Reply:
I'm pretty sure if an item is already in the set it just doesn't get added, therefore removing it first ensures that it gets overwritten.

Reply:
Backing up a little bit: why are we editing the NameMultimap class at all? As the name suggests, duplicates should be always be allowed.

Reply:
From my knowledge NameMultimap is used for functions because it can store overloads of a function (which are not duplicates in the sense that I am defining them to be). If we used NameMap, we would not be able to support overloads.

Reply:
Ah I see, so Multimap just refers to allowing multiple functions with the same name but different parameter lengths?

This seems like a fair amount of changes to add support for duplicate functions (as in same name and number of parameters). Would it be simpler just to throw an error if a duplicate function is detected rather than allowing for it to happen?

Reply:
From a glance the amount of changes seem similar since allowing duplicates is the default behavior so the custom boolean would still be required. However, we could avoid needing to make the List a collection probably. My main concern is that I don't think that the schema resets during tests so some additional work would need to go into exposing a reset mechanism (this may be trivial but I haven't take a look yet).

Reply:
Based on the test in UtilTest::testNameMultimap, it appears that the intention of this class is to explicitly allow duplicates of whatever object it is storing. By making these proposed changes, it feels like we're kind of abusing this original intent.

What if we instead stored CalciteSchema's FunctionEntry objects in a NameMap? And we updated the hashCode() and equals() methods for the FunctionEntryImpl (or something similar) to be able to differentiate two functions with the same name but different argument lengths? Then we would automatically get deduplication for free from that data structure.

Reply:
Is this what you had in mind? https://github.com/AndrewPochapsky/calcite/compare/function-schema...AndrewPochapsky:name-map?expand=1

To me this approach looks equivalent to using a NameMultimap minus the abstraction.